src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs

```diff
@@ -390,28 +495,25 @@ internal static void Unescape(ReadOnlySpan<byte> source, Span<byte> destination,
 
                             Debug.Assert(JsonHelpers.IsInRangeInclusive((uint)scalar, JsonConstants.HighSurrogateStartValue, JsonConstants.HighSurrogateEndValue));
 
-                            idx += 3;   // Skip the last hex digit and the next \u
-
                             // We must have a low surrogate following a high surrogate.
-                            if (source.Length < idx + 4 || source[idx - 2] != '\\' || source[idx - 1] != 'u')
+                            if (source.Length < idx + 7 || source[idx + 1] != '\\' || source[idx + 2] != 'u')
                             {
-                                ThrowHelper.ThrowInvalidOperationException_ReadInvalidUTF16();
+                                ThrowHelper.ThrowInvalidOperationException_ReadIncompleteUTF16();
                             }
 
                             // The source is known to be valid JSON, and hence if we see a \u, it is guaranteed to have 4 hex digits following it
                             // Otherwise, the Utf8JsonReader would have already thrown an exception.
-                            result = Utf8Parser.TryParse(source.Slice(idx, 4), out int lowSurrogate, out bytesConsumed, 'x');
+                            result = Utf8Parser.TryParse(source.Slice(idx + 3, 4), out int lowSurrogate, out bytesConsumed, 'x');
```

Comment:
I personally prefer when we start slicing with `idx` so I'd rather have ass increment 3 first and then another 3. Does this make any perf difference?

Reply:
I did this originally, but it appears that doing slicing all too aggressively does impact performance substantially.